src/qt/textctrl.cpp

```diff
@@ -14,19 +14,364 @@
 #include "wx/qt/private/winevent.h"
 #include "wx/qt/private/utils.h"
 
+#include <algorithm>
 #include <QtWidgets/QLineEdit>
 #include <QtWidgets/QTextEdit>
 
+
+namespace
+{
 class wxQtLineEdit : public wxQtEventSignalHandler< QLineEdit, wxTextCtrl >
 {
 public:
     wxQtLineEdit( wxWindow *parent, wxTextCtrl *handler );
 
 private:
-    void textChanged(const QString &text);
+    void textChanged();
     void returnPressed();
 };
 
+class wxQtTextEdit : public wxQtEventSignalHandler< QTextEdit, wxTextCtrl >
+{
+public:
+    wxQtTextEdit( wxWindow *parent, wxTextCtrl *handler );
+
+private:
+    void textChanged();
+};
+
+class wxQtMultiLineEdit : public wxQtEdit
+{
+public:
+    wxQtMultiLineEdit(QTextEdit *edit) : edit(edit)
+    {
+    }
+
+    virtual bool IsModified() const wxOVERRIDE
+    {
+        return edit->isWindowModified();
+    }
+
+    virtual wxString DoGetValue() const wxOVERRIDE
+    {
+        return wxQtConvertString( edit->toPlainText() );
+    }
+
+    virtual long GetInsertionPoint() const wxOVERRIDE
+    {
+        QTextCursor cursor = edit->textCursor();
+        return cursor.anchor();
+    }
+
+    virtual QWidget *GetHandle() const wxOVERRIDE
+    {
+        return edit;
+    }
+
+    virtual int GetNumberOfLines() const wxOVERRIDE
+    {
+        const wxString &value = DoGetValue();
+        return std::count(value.begin(), value.end(), '\n') + 1;
+    }
+
+    std::pair<size_t, size_t> getLineStart(long lineNo, const wxString &value) const
+    {
+        size_t pos = 0;
+        long cnt = 0;
+
+        while(cnt < lineNo)
+        {
+            size_t tpos = value.find('\n', pos);
+            if(tpos == wxString::npos) return std::make_pair(tpos, tpos);
```

Comment:
Minor style comment: please don't put compound statement body on the same line as the statement itself.